### DISCUSSION TITLE
How does iperf determine packet loss?

### DISCUSSION THREAD

**Discussion by fyyxxm** [esnet/iperf#1680] [Q&A]
Hello, I am conducting a test on the network card's packet transmission and reception using iperf3.  While running the test, iperf indicated that there were packet losses, but when we checked the number of packets received on the network card driver, it was accurate.

I would like to know if the indication of packet loss by iperf can also be due to packet errors.  If so, what is the current criterion used by iperf to determine packet errors?

**Comment by bmah888** [comment]
Yes, it's possible for packets to get dropped on the receiving host, for example if there's not enough socket buffer space to hold the incoming data.

Are you doing tests with TCP, UDP, or SCTP? If TCP, iperf3 doesn't explicitly track packet loss, but (at least on Linux) it does read the number of TCP retransmissions from the `tcpinfo` structure in the kernel and report that out. For UDP, the sending iperf3 puts a sequence number in every UDP datagram, the receiving iperf3 can then count losses by detecting gaps in the sequence numbers of received packets. (Note this doesn't quite work right in the case of packet reordering or duplication.) I'm not sure how we report on losses with SCTP but chances are you're not using that anyways.

Hope this helps!

**Comment by fyyxxm** [comment]
Yes, we used UDP for testing, I will test again to determine whether the sequence number error caused the perf message packet loss.
Thank you for your help
